### PR TITLE
[SPARK-22188] [CORE] Adding security headers for preventing XSS, MitM and MIME sniffing

### PR DESCRIPTION
**Pull request by krishna-pandey**
## What changes were proposed in this pull request?

The HTTP Strict-Transport-Security response header (often abbreviated as HSTS) is a security feature that lets a web site tell browsers that it should only be communicated with using HTTPS, instead of using HTTP.

Note: The Strict-Transport-Security header is ignored by the browser when your site is accessed using HTTP; this is because an attacker may intercept HTTP connections and inject the header or remove it. When your site is accessed over HTTPS with no certificate errors, the browser knows your site is HTTPS capable and will honor the Strict-Transport-Security header.

The HTTP X-XSS-Protection response header is a feature of Internet Explorer, Chrome and Safari that stops pages from loading when they detect reflected cross-site scripting (XSS) attacks.

The HTTP X-Content-Type-Options response header is used to protect against MIME sniffing vulnerabilities.

## How was this patch tested?
Checked on my system locally.

screen shot 2017-10-03 at 6 49 20 pm

### CODE REVIEW

File: conf/spark-defaults.conf.template [hunk 1]
```diff
@@ -19,9 +19,16 @@
...
-# spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+# spark.master                          spark://master:7077
+# spark.eventLog.enabled                true
+# spark.eventLog.dir                    hdfs://namenode:8021/directory
+# spark.serializer                      org.apache.spark.serializer.KryoSerializer
+# spark.driver.memory                   5g
+# spark.executor.extraJavaOptions       -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+
+#spark.ui.allowFramingFrom              https://example.com/
+#spark.ui.xXssProtection.enabled        1; mode=block
```
Comment (srowen) [inline conf/spark-defaults.conf.template]: This sounds like a boolean flag but its value isn't boolean; if it has only one reasonable value then it can be a boolean that sets that value. Otherwise maybe call the property something else

File: conf/spark-defaults.conf.template [hunk 2]
```diff
@@ -19,9 +19,16 @@
...
+#spark.ui.xContentType.options          nosniff
+
+#Enable below only when Spark is running on HTTPS
+#spark.ui.strictTransportSecurity.age   max-age=31536000
```
Comment (srowen) [inline conf/spark-defaults.conf.template]: On a similar note does max-age need to be part of the user-supplied string or just the age? in what units?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 3]
```diff
@@ -89,6 +92,9 @@ private[spark] object JettyUtils extends Logging {
             val result = servletParams.responder(request)
             response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
             response.setHeader("X-Frame-Options", xFrameOptionsValue)
+            response.setHeader("X-XSS-Protection", xXssProtectionValue.get)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: This will fail if these aren't set. Only set the headers if the Option exists

**Review by srowen — CHANGES_REQUESTED** [review]
Seems reasonable as a sort of advanced developer option

File: conf/spark-defaults.conf.template (hunk 1 above)
Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: Possible values can be one from the below set, default one mentioned in template is commonly-used and provides effective protection.
X-XSS-Protection: { 0 | 1 |1; mode=block | 1; report=<reporting-uri> }

File: conf/spark-defaults.conf.template (hunk 2 above)
Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: Yes, "max-age" need to be part of user-supplied string. Possible values can be one from the below set, default one mentioned in template is for optimum secure value commonly used.

Strict-Transport-Security: {max-age=\<expire-time\> | max-age=\<expire-time\>; includeSubDomains | max-age=\<expire-time\>; preload}


Value is in delta-seconds. More here https://tools.ietf.org/html/rfc6797#section-6.1.1

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 3 above)
Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Yes, I will add a check for that.

File: conf/spark-defaults.conf.template (hunk 1 above)
Comment (srowen) [reply conf/spark-defaults.conf.template]: I would remove '.enabled' from the name then

File: conf/spark-defaults.conf.template (hunk 2 above)
Comment (srowen) [reply conf/spark-defaults.conf.template]: I'd just remove '.age' then, likewise. It's the very value of the STS header

Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: I understand that the "max-age" part is common across all values but don't want to tamper the value part for ease and any future compatibility. I will rather remove ".age" to avoid confusion arising out of this.

File: conf/spark-defaults.conf.template (hunk 1 above)
Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: @srowen renamed the keys as suggested. Thanks again for the review.

File: conf/spark-defaults.conf.template [hunk 4]
```diff
@@ -19,9 +19,16 @@
...
-# spark.eventLog.dir               hdfs://namenode:8021/directory
-# spark.serializer                 org.apache.spark.serializer.KryoSerializer
-# spark.driver.memory              5g
-# spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+# spark.master                          spark://master:7077
+# spark.eventLog.enabled                true
+# spark.eventLog.dir                    hdfs://namenode:8021/directory
+# spark.serializer                      org.apache.spark.serializer.KryoSerializer
+# spark.driver.memory                   5g
+# spark.executor.extraJavaOptions       -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
```
Comment (dongjoon-hyun) [inline conf/spark-defaults.conf.template]: Hi, @krishna-pandey .
It seems that you can revert the change on line 22 ~ 27.
Could you re-adjust the spacing?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 5]
```diff
@@ -89,6 +92,15 @@ private[spark] object JettyUtils extends Logging {
...
             response.setHeader("X-Frame-Options", xFrameOptionsValue)
+            if (xXssProtectionValue.isDefined) {
+              response.setHeader("X-XSS-Protection", xXssProtectionValue.get)
+            }
+            if (xContentTypeOptionsValue.isDefined) {
+              response.setHeader("X-Content-Type-Options", xContentTypeOptionsValue.get)
+            }
+            if (strictTransportSecurityValue.isDefined) {
+              response.setHeader("Strict-Transport-Security", strictTransportSecurityValue.get)
+            }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Can we simplify line 95~103 into the following?
```scala
xXssProtectionValue.foreach(response.setHeader("X-XSS-Protection", _))
xContentTypeOptionsValue.foreach(response.setHeader("X-Content-Type-Options", _))
strictTransportSecurityValue.foreach(response.setHeader("Strict-Transport-Security", _))
```

**Comment by dongjoon-hyun** [comment]
I can see them from the browser.
LGTM except two minor comments, @krishna-pandey .

**Comment by krishna-pandey** [comment]
@dongjoon-hyun Thanks for the review. Made the changes as suggested.

File: conf/spark-defaults.conf.template [hunk 6]
```diff
@@ -25,3 +25,10 @@
 # spark.serializer                 org.apache.spark.serializer.KryoSerializer
 # spark.driver.memory              5g
 # spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+
+#spark.ui.allowFramingFrom         https://www.example.com/
+#spark.ui.xXssProtection           1; mode=block
+#spark.ui.xContentType.options     nosniff
+
+#Enable below only when Spark is running on HTTPS
+#spark.ui.strictTransportSecurity  max-age=31536000
```
Comment (dongjoon-hyun) [inline conf/spark-defaults.conf.template]: Could you add a space after `#` and an empty line at the end of file?
```
# spark.ui.allowFramingFrom         https://www.example.com/
# spark.ui.xXssProtection           1; mode=block
# spark.ui.xContentType.options     nosniff

# Enable below only when Spark is running on HTTPS
# spark.ui.strictTransportSecurity  max-age=31536000
```

**Comment by krishna-pandey** [comment]
@dongjoon-hyun Made the changes as suggested.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM.

File: conf/spark-defaults.conf.template [hunk 7]
```diff
@@ -25,3 +25,10 @@
 # spark.serializer                 org.apache.spark.serializer.KryoSerializer
 # spark.driver.memory              5g
 # spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+
+# spark.ui.allowFramingFrom         https://www.example.com/
```
Comment (srowen) [inline conf/spark-defaults.conf.template]: Hm, my last thought on this is that these are undocumented options, and maybe rightly so as they're kind of advanced options. But if so they don't really have a place in the default props template; what do they mean to a user?

I think they should actually be removed here, on second glance, but, wouldn't be a big deal to document them in the configuration doc as they're arguably non-internal, important functionality for some users.

Comment (jerryshao) [reply conf/spark-defaults.conf.template]: Agree with @srowen , we should remove the configurations here in template, since they're not common configurations. Also add them to `docs/configuration.md`.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 8]
```diff
@@ -79,6 +79,9 @@ private[spark] object JettyUtils extends Logging {
     val allowFramingFrom = conf.getOption("spark.ui.allowFramingFrom")
     val xFrameOptionsValue =
       allowFramingFrom.map(uri => s"ALLOW-FROM $uri").getOrElse("SAMEORIGIN")
+    val xXssProtectionValue = conf.getOption("spark.ui.xXssProtection")
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Please use `ConfigEntry` for newly added configurations, you could refer to `org.apache.spark.internal.config`.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 9]
```diff
@@ -89,6 +92,9 @@ private[spark] object JettyUtils extends Logging {
             val result = servletParams.responder(request)
             response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
             response.setHeader("X-Frame-Options", xFrameOptionsValue)
+            xXssProtectionValue.foreach(response.setHeader("X-XSS-Protection", _))
+            xContentTypeOptionsValue.foreach(response.setHeader("X-Content-Type-Options", _))
+            strictTransportSecurityValue.foreach(response.setHeader("Strict-Transport-Security", _))
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: The changes here will also affect HTTP request, is that OK?

Also is it enough to only change the GET request?

Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: @jerryshao Out of these three, Strict Transport Security header makes sense, when SSL/TLS is enabled.

**Comment by jerryshao** [comment]
@vanzin @tgravescs @ajbozarth  what is your opinion on this PR? Is it a necessary fix for Spark?

File: conf/spark-defaults.conf.template [hunk 10]
```diff
@@ -25,3 +25,10 @@
...
+# spark.ui.xXssProtection           1; mode=block
+# spark.ui.xContentType.options     nosniff
+
+# Enable below only when Spark is running on HTTPS
+# spark.ui.strictTransportSecurity  max-age=31536000
```
Comment (jerryshao) [inline conf/spark-defaults.conf.template]: What's the meaning of this specific number "31536000"?

Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: The REQUIRED "max-age" directive specifies the number of seconds, after the reception of the STS header field, during which the UA regards the host (from whom the message was received) as a Known HSTS Host. Here the value is equal to 365 days. More at https://tools.ietf.org/html/rfc6797#section-6.1.1

**Comment by srowen** [comment]
I think the change is OK, to give a mechanism to set these headers if desired. There are still several comments to address though.

File: conf/spark-defaults.conf.template (hunk 7 above)
Comment (krishna-pandey) [reply conf/spark-defaults.conf.template]: @srowen , @jerryshao Removed the entries from the config template and put it to configuration doc as suggested.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 8 above)
Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: @jerryshao I am not sure if that will add much value in this context apart from following best practices.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 9 above)
Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: @jerryshao I added a check to set the STS header, only if SSL is enabled. However, STS header was not stopping Browser from rendering the content even when it was set and Spark was running on plain HTTP. Need to get expected behaviour here.
screen shot 2017-10-12 at 6 22 18 pm

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 11]
```diff
@@ -89,6 +92,13 @@ private[spark] object JettyUtils extends Logging {
             val result = servletParams.responder(request)
             response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
             response.setHeader("X-Frame-Options", xFrameOptionsValue)
+            xXssProtectionValue.foreach(response.setHeader("X-XSS-Protection", _))
+            if (xContentTypeOptionsValue.get.equalsIgnoreCase("true")) {
+              response.setHeader("X-Content-Type-Options", "nosniff")
+            }
+            if (conf.get("spark.ssl.enabled").equalsIgnoreCase("true")) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: I think you can check `request.scheme` if it is "https" or not?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 8 above)
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: It follows new code convention, newly added configurations is better to change to that way.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 11 above)
Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: @jerryshao Thanks for the tip. I will do that.

**Comment by tgravescs** [comment]
overall I think the headers are fine, more security the better.  I'm not exactly sure the attack vector with the spark UI though.  Normally I would expect your UI to be on a corporate network and you vpn in, but I guess maybe if you are running in AWS or similar public cloud and you go somewhere to access, but I'm not sure what data they can get there or why you would be using http in the first place but there are lots of setups.  

@krishna-pandey  do you have specific use case/attack vector in mind here?

I was wondering if there was a more generic way to allow user to specify desired headers without having a config for each one. Downside to that is its not as obvious though too so I need to think about that a bit more.

**Comment by krishna-pandey** [comment]
@tgravescs These generic headers are about providing available client-side protection for the application. I also think even if there is no sensitive data to formulate an attack by itself here, the information can be used in conjunction to target other ecosystem components. Also, in future we may add an interface for data access. Now is the time to think of Security First. Cross-site Scripting is one of the most prevalent attack vector and has been an OWASP Top 10 risk for web applications for decades. As the effort to have these in place here is minimal, IMHO we should set these. 

As you rightly mentioned, deployment on cloud can expand the attack surface pretty wide in absence of right firewall policy. Also let's not forget insider threat inside corporate networks.

Going forward may be we will have enough insight to choose which headers are needed to be enabled by default and enforce them from application side and not leave it to Users.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 12]
```diff
@@ -420,6 +420,25 @@ package object config {
...
+  private[spark] val UI_X_XSS_PROTECTION =
+    ConfigBuilder("spark.ui.xXssProtection")
+      .doc("Value for HTTP X-XSS-Protection response header")
+      .stringConf
+      .createOptional
+
+  private[spark] val UI_X_CONTENT_TYPE_OPTIONS =
+    ConfigBuilder("spark.ui.xContentTypeOptions.enabled")
+      .doc("Set to 'true' for setting X-Content-Type-Options HTTP response header to 'nosniff'")
+      .stringConf
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: This should be a boolean option then?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 13]
```diff
@@ -89,6 +90,14 @@ private[spark] object JettyUtils extends Logging {
             val result = servletParams.responder(request)
             response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
             response.setHeader("X-Frame-Options", xFrameOptionsValue)
+            conf.get(UI_X_XSS_PROTECTION).foreach(response.setHeader("X-XSS-Protection", _))
+            if (conf.get(UI_X_CONTENT_TYPE_OPTIONS.key).equals("true")) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: At the moment this would fail if unset

File: docs/configuration.md [hunk 14]
```diff
@@ -2013,7 +2013,62 @@ Apart from these, the following properties are also available, and may be useful
...
+<table class="table">
+    <tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
+    <tr>
+        <td><code>spark.ui.xXssProtection</code></td>
+        <td>None</td>
+        <td>
+            Value for HTTP X-XSS-Protection response header. You can choose appropriate value 
+            from below:
+            <ul>
+                <li>  0 (Disables XSS filtering)
```
Comment (srowen) [inline docs/configuration.md]: Markdown won't work here? I think it won't. Close the `<li>` tags though. It's not clear below what the values the users sets though. The values are `0`, `1`, or `1; mode=block` and that should be clear.

File: docs/configuration.md [hunk 15]
```diff
@@ -2013,7 +2013,62 @@ Apart from these, the following properties are also available, and may be useful
...
+        </td>
+    </tr>
+    <tr>
+        <td><code>spark.ui.allowFramingFrom</code></td>
+        <td>SAMEORIGIN</td>
+        <td>
+            Value for X-Frame-Options HTTP response header
+            <br />You can provide the "website uri" which can only be displayed in a frame on 
+                the specified origin. 
+            <br />
```
Comment (srowen) [inline docs/configuration.md]: Remove this

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/configuration.md [hunk 16]
```diff
@@ -2013,7 +2013,62 @@ Apart from these, the following properties are also available, and may be useful
     </tr>
 </table>
 
+### HTTP Security Headers
```
Comment (jerryshao) [inline docs/configuration.md]: I think this could move to `security.md` as a security related advanced configurations.

File: docs/configuration.md (hunk 14 above)
Comment (krishna-pandey) [reply docs/configuration.md]: Removed the lists, however it is being used in many places in existing configuration.md

File: docs/configuration.md (hunk 15 above)
Comment (krishna-pandey) [reply docs/configuration.md]: Done.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 12 above)
Comment (krishna-pandey) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Making it to Boolean.

**Comment by jerryshao** [comment]
ok to test.

**Comment by krishna-pandey** [comment]
@srowen @tgravescs @dongjoon-hyun @jerryshao 
Please review the PR. I have incorporated all changes as suggested.

**Comment by jerryshao** [comment]
>/home/jenkins/workspace/SparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/internal/config/package.scala:440:0: Whitespace at end of line

Please fix the style issue.

**Comment by krishna-pandey** [comment]
@jerryshao removed Whitespace at end of line 440 in package.scala. ok to test.

File: docs/security.md [hunk 17]
```diff
@@ -186,7 +186,52 @@ configure those ports.
...
+
+<table class="table">
+    <tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
+    <tr>
+        <td><code>spark.ui.xXssProtection</code></td>
+        <td>None</td>
+        <td>
+            Value for HTTP X-XSS-Protection response header. You can choose appropriate value 
+            from below:
+            <br />
```
Comment (srowen) [inline docs/security.md]: Why not just leave this as a bulleted list? Not a big deal I guess just less conventional for HTML

File: docs/configuration.md [hunk 18]
```diff
@@ -2013,7 +2013,6 @@ Apart from these, the following properties are also available, and may be useful
     </tr>
 </table>
 
-
```
Comment (srowen) [inline docs/configuration.md]: If you have to change the pull request again, I'd revert this, but no need to change it only for this

File: docs/security.md (hunk 17 above)
Comment (krishna-pandey) [reply docs/security.md]: @srowen Made them bulleted list and closed the <li> tag as well.

File: docs/configuration.md (hunk 18 above)
Comment (krishna-pandey) [reply docs/configuration.md]: Restored the change.

File: docs/security.md [hunk 19]
```diff
@@ -186,7 +186,54 @@ configure those ports.
...
 </table>
 
+### HTTP Security Headers
+
+Apache Spark can be configured to include HTTP Headers which aids in preventing Cross 
+Site Scripting (XSS), Cross-Frame Scripting (XFS), MIME-Sniffing and also enforces HTTP 
+Strict Transport Security.
+
+<table class="table">
+    <tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
```
Comment (jerryshao) [inline docs/security.md]: I think in Spark we follow 2 space indent for html code. You could refer to other docs.

Comment (krishna-pandey) [reply docs/security.md]: @jerryshao Fixed the indentation issue as per other Spark docs. Thanks for pointing that out.

**Comment by jerryshao** [comment]
LGTM.

File: docs/security.md [hunk 20]
```diff
@@ -186,7 +186,54 @@ configure those ports.
...
+<table class="table">
+<tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
+<tr>
+  <td><code>spark.ui.xXssProtection</code></td>
+  <td>None</td>
+  <td>
+    Value for HTTP X-XSS-Protection response header. You can choose appropriate value 
+    from below:
+    <ul>
+      <li> 0 </li> (Disables XSS filtering)
```
Comment (srowen) [inline docs/security.md]: Now you have text outside the list item. This won't render correctly. This should be for example:
```
<li><code>0</code> (Disables XSS filtering)</li>
```

Comment (krishna-pandey) [reply docs/security.md]: @srowen Thanks for the excellent tip. It looks more readable now.

**Comment by rxin** [comment]
Is there a reason why this cannot be always enabled?

**Comment by srowen** [comment]
The downside of additional header traffic is trivial. I suppose it is not needed in most contexts so it was conservative to leave it off by default. That said I am not aware of any particular negative to enabling it. I would be OK with leaving it on by default if there's no particular downside.

**Comment by rxin** [comment]
Yea in general for security features it seems like it's good to turn on them by default.

**Comment by krishna-pandey** [comment]
@rxin , @srowen I think we can enable X-XSS-Protection and X-Content-Type-Options response header by default. STS Header can be left configurable or enabled by default when Spark UI is running on HTTPS.

**Word of caution**: When X-Content-Type-Options response HTTP header is set to "nosniff", it will block a request if the requested type is  "style" and the MIME type is not "text/css", or when requested type is "script" and the MIME type is not a JavaScript MIME type.

**Comment by srowen** [comment]
For the content type header, does the UI seem to work normally with it on? I don't think we should have any problem with that restriction or else need to fix it .  You could move the defaults to on then.

**Comment by krishna-pandey** [comment]
@srowen @rxin Made changes to enable the X-Content-Type-Options and X-XSS-Protection values by default. Please review.

**Comment by srowen** [comment]
Merged to master
